### PR TITLE
refactor: ChromaDB migration

### LINKED ISSUE #19: Feature: Explore chromadb
### Type of feature

🍕 Feature

### Current behavior

Thanks to @Anush008 there is a [chromadb rust](https://docs.rs/chromadb/latest/chromadb/) client. We should make plans to explore this in the repo-query and evaluate if there is anything missing in the implementation. 

### Suggested solution

Let's work on a branch and see if any functionality is lost. We can leverage their [community](https://discord.gg/MMeYNTmh3x) for support too. 

### Additional context

https://github.com/Anush008/chromadb-rs

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### Contributing Docs

- [x] I agree to follow this project's Contribution Docs

### LINKED ISSUE #21: Collection example in the README seems off
Should this example: https://github.com/open-sauced/repo-query#3-get-collection be /collection instead?

### PR DESCRIPTION
**Pull request by Anush008**
## Description
* This PR intends to migrate the project to use ChromaDB by implementing the `RepositoryEmbeddingsDB` trait for the ChromaDB client.
https://github.com/open-sauced/repo-query/blob/117a93398a6c771f004d328ac1f39ea546c6ad81/src/db/mod.rs#L9-L22

* The README has been updated with the new local development steps

* The ChromaDB service has been added to the `docker-compose.yaml` file.

## What type of PR is this? (check all applicable)

- [x] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [x] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [x] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

## Related Tickets & Documents
Resolves #19 

## Added tests?

- [ ] 👍 yes
- [x] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

## Added to documentation?

- [x] 📜 README.md
- [ ] 📓 docs.opensauced.pizza
- [ ] 🍕 dev.to/opensauced
- [ ] 📕 storybook
- [ ] 🙅 no documentation needed

### CODE REVIEW

**Comment by Anush008** [comment]
@bdougie, @jpmcb, you can try the service now.
Everything should stay the same.
You can set the `OPENAI_API_KEY` and `WEBSERVER_PORT` values in a `.env` file and run `make up` quickly get this running.

**Comment by jpmcb** [comment]
Wow excellent work. I'll review this today and look at getting this deployed on our azure instance. I have some terraform that should make that a quick process.

**Comment by Anush008** [comment]
Hey @jpmcb. The `ort` lib supports some execution providers that can improve the embedding performance.
https://github.com/pykeio/ort#execution-providers

We are currently using the CPU provider. https://github.com/open-sauced/repo-query/blob/afc4d19068e7c84a2566dae9598f1500f1191705/src/embeddings/onnx.rs#L22-L24.

Moving forward, if needed, we can try some of the others if the hardware support is available.

**Comment by jpmcb** [comment]
> if the hardware support is available

GPU support would be cool, but right now, for the way we have it deployed through Container Apps, there aren't GPU instances available. This is essentially their serverless container runtime and doesn't involve alot of infrastructure. We'd need to look at a more intensive deployment strategy, maybe through Container Instances or Azure's Kubernetes service, that would let us deploy GPU instances.

For now, until we need to start scaling up in that way, I think the Container Apps deployment with CPU `ort` optimization is just fine.

**Review by bdougie — APPROVED** [review]
Looks like a pretty clean swap. Really appreciate you kicking off the chroma client work as well.

File: src/db/mod.rs [hunk 1]
```diff
@@ -1,23 +1,22 @@
...
         repository: &Repository,
         query_embeddings: Embeddings,
         limit: usize,
     ) -> Result<RepositoryFilePaths>;
 
-    async fn get_file_paths(&self, repository: &Repository) -> Result<RepositoryFilePaths>;
+    fn get_file_paths(&self, repository: &Repository) -> Result<RepositoryFilePaths>;
 
-    async fn is_indexed(&self, repository: &Repository) -> Result<bool>;
+    fn is_indexed(&self, repository: &Repository) -> Result<bool>;
```
Comment (jpmcb) [inline src/db/mod.rs:21]: Makes sense we'd remove `async` since before we were just `await` on each call anyways 👍🏼

File: .env.example [hunk 2]
```diff
@@ -1,4 +1,4 @@
-OPENAI_API_KEY=
-QDRANT_URL=         #Defaults to http://localhost:6334
-RUST_LOG=           #Logging levels: "error", "warn", "info", "debug", "trace"
-WEBSERVER_PORT=     #Defaults to 3000
\ No newline at end of file
+OPENAI_API_KEY=     #REQUIRED
+WEBSERVER_PORT=     #REQUIRED
+CHROMA_URL=         #Defaults to http://localhost:8000
```
Comment (jpmcb) [inline .env.example:3]: Where does this get used to set the Chroma client? I'm not seeing any other references to this in the code. Is it being consumed by the Chroma rust client?

**Review by jpmcb — COMMENTED** [review]
I was having some snags running the image locally on arm64 so i opened: https://github.com/chroma-core/chroma/issues/972

But I was able to rebuild the image and this works well - nice work!

Performance seems slightly worse than qdrant locally for me. But will have to try that on real hardware.

File: .env.example (hunk 2 above)
Comment (Anush008) [reply .env.example:3]: Yes. The Chromadb-rs lib uses the `CHROMA_URL` env  by default.
https://github.com/open-sauced/repo-query/blob/37bd46815bc9eb765bb9a271783023c1f1ad9c5c/src/db/chroma.rs#L21

**Comment by Anush008** [comment]
https://github.com/open-sauced/repo-query/pull/20/commits/cc7f60612d1e11c7baa9bbd618c7109df45d8325, resolves #21.

**Review by jpmcb — CHANGES_REQUESTED** [review]
There are a few non-starters to getting this merged:

- https://github.com/chroma-core/chroma/issues/1002
  - Rebuilding during container entry causes quite a lengthy gap for the repo-query service to be ready. Ideally, since we're deploying all this in a conainer app environment, we'd not have to wait for container bits to be rebuilt
  - https://github.com/chroma-core/chroma/issues/993
    - If we had a precompiled form of the chroma database that we could drop into our own container, that would alleviate this pain, so we may want to watch that issue as well
- https://github.com/chroma-core/chroma/issues/972
  - The team is working on multi-arch builds and that will help alot with local development for our contributor community

Until then, we should hold this migration and stick with Qdrant since it's working out of the box in our container environments.

**Pull request closed without merging**